File: bot.py
import os
import time
import sqlite3
from collections import defaultdict
import discord
from discord.ext import commands

from keep_alive import keep_alive

DB_NAME = "database.db"

def init_db():
    conn = sqlite3.connect(DB_NAME)
    cursor = conn.cursor()
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS users (
            user_id INTEGER PRIMARY KEY,
            username TEXT,
            joined_at TEXT,
            warnings INTEGER DEFAULT 0
        )
    """)
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS user_roles (
            user_id INTEGER,
            role_id INTEGER,
            PRIMARY KEY (user_id, role_id)
        )
    """)
    conn.commit()
    conn.close()
    print("[DATABÁZE] SQLite databáze byla úspěšně inicializována.")

init_db()

class RoleSelectView(discord.ui.View):
    def __init__(self):
        super().__init__(timeout=None)

    async def toggle_role(self, interaction: discord.Interaction, role_id_str: str):
        if not role_id_str.isdigit():
            await interaction.followup.send("⚠️ Zadané ID role není číslo.", ephemeral=True)
            return

        role_id = int(role_id_str)
        role = interaction.guild.get_role(role_id)
        if not role:
            await interaction.followup.send(f"⚠️ Role s ID `{role_id_str}` nebyla na serveru nalezena.", ephemeral=True)
            return

        conn = sqlite3.connect(DB_NAME)
        cursor = conn.cursor()

        try:
            if role in interaction.user.roles:
                await interaction.user.remove_roles(role)
                cursor.execute("DELETE FROM user_roles WHERE user_id = ? AND role_id = ?", (interaction.user.id, role_id))
                await interaction.followup.send(f"❌ Role **{role.name}** ti byla odebrána.", ephemeral=True)
            else:
                await interaction.user.add_roles(role)
                cursor.execute("INSERT OR IGNORE INTO user_roles (user_id, role_id) VALUES (?, ?)", (interaction.user.id, role_id))
                await interaction.followup.send(f"✅ Role **{role.name}** ti byla přidána!", ephemeral=True)
            
            conn.commit()
        except discord.Forbidden:
            await interaction.followup.send("❌ **Chyba Oprávnění:** Bot nemá práva upravit tuto roli!", ephemeral=True)
        except Exception as e:
            await interaction.followup.send(f"⚠️ Nepodařilo se upravit roli: {e}", ephemeral=True)
        finally:
            conn.close()

    @discord.ui.select(
        placeholder="🎨 Vyber si barvu jména...",
        custom_id="select_color_role",
        options=[
            discord.SelectOption(label="Červená", value="1540028254885257256", emoji="🔴"),
            discord.SelectOption(label="Modrá", value="1540028432614817862", emoji="🔵"),
            discord.SelectOption(label="Zelená", value="1540028872312225914", emoji="🟢"),
            discord.SelectOption(label="Fialová", value="1540028968273449002", emoji="🟣"),
            discord.SelectOption(label="Žlutá", value="1553041371986927657", emoji="🟡"),
        ]
    )
    async def color_select_callback(self, interaction: discord.Interaction, select: discord.ui.Select):
        await interaction.response.defer(ephemeral=True)
        await self.toggle_role(interaction, select.values[0])

    @discord.ui.select(
        placeholder="🎂 Vyber si svůj věk...",
        custom_id="select_age_role",
        options=[
            discord.SelectOption(label="13-17+", value="1553032262696566915", emoji="🔞"),
            discord.SelectOption(label="18+", value="1553032409073586186", emoji="🔞"),
        ]
    )
    async def age_select_callback(self, interaction: discord.Interaction, select: discord.ui.Select):
        await interaction.response.defer(ephemeral=True)
        await self.toggle_role(interaction, select.values[0])

class DiscordBot(commands.Bot):
    def __init__(self):
        intents = discord.Intents.default()
        intents.message_content = True
        intents.members = True
        super().__init__(command_prefix="!", intents=intents)
        self.user_message_timestamps = defaultdict(list)

    async def setup_hook(self):
        self.add_view(RoleSelectView())

    async def on_ready(self):
        print(f"[BOT] Přihlášen jako: {self.user} (ID: {self.user.id})")

    async def on_message(self, message):
        if message.author.bot or not message.guild:
            return

        if message.author.guild_permissions.administrator:
            await self.process_commands(message)
            return

        if any(link in message.content.lower() for link in ["http://", "https://", "discord.gg/"]):
            try:
                await message.delete()
                await message.channel.send(f"{message.author.mention}, posílání odkazů je zakázáno!", delete_after=5)
                return
            except Exception:
                pass

        now = time.time()
        user_timestamps = self.user_message_timestamps[message.author.id]
        user_timestamps.append(now)
        self.user_message_timestamps[message.author.id] = [t for t in user_timestamps if now - t <= 5]

        if len(self.user_message_timestamps[message.author.id]) > 5:
            try:
                await message.delete()
                await message.channel.send(f"{message.author.mention}, nepřestávej spamovat!", delete_after=5)
                return
            except Exception:
                pass

        await self.process_commands(message)

bot = DiscordBot()

@bot.command(name="setup_roles")
@commands.has_permissions(administrator=True)
async def setup_roles(ctx):
    msg_text = """# role

<@&1539241329899474954> - **je vlastník serveru který rozhoduje co se přidá na server**
<@&1552268147179135046> - **je člen A-Teamu který dělá s majitelem změny na serveru a hlídá server**
<@&1539240450810978364> - **je člen A-Teamu který taky může změny na serveru ale na žádost majitele a hlídá server**
<@&1552391311754133555> - **je člen A-Teamu který testuje jestli všechno funguje**
<@&1549866576029814784> - **je role pro členy který jsou aktivní na serveru a na vc**
<@&1539242304013992048> - **je role který maji všechny na serveru**
<@&1543385078992871524> / <@&1543385813943984239> - **je pro členy který si požádaji o roli**
<@&1540028254885257256> / <@&1540028432614817862> / <@&1540028872312225914> / <@&1540028968273449002> / <@&1553041371986927657> - **je členy který si požádaji o barvu kterou chci na nicku na serveru**
<@&1553032262696566915> / <@&1553032409073586186> - **je členy který si požádaji o věk který chtějí mít na serveru**
<@&1549872649877069924> - **je pro členy který dají booster na server**

👇 **Vyber si své role v menu níže:**"""
    
    await ctx.send(content=msg_text, view=RoleSelectView())
    await ctx.message.delete()

if __name__ == "__main__":
    keep_alive()
    token = os.getenv("DISCORD_TOKEN")
    if token:
        bot.run(token)
    else:
        print("[CHYBA] Proměnná DISCORD_TOKEN nebyla nalezena!")

File: bot.js
import { logger } from '../utils/logger.js';

export const botConfig = {
  // =========================
  // BOT PRESENCE (what users see under the bot name)
  // =========================
  // `status` options:
  // - "online"    = green dot
  // - "idle"      = yellow moon
  // - "dnd"       = red do-not-disturb
  // - "invisible" = appears offline
  presence: {
    // Current online state shown on Discord.
    status: "online",

    // Activity lines shown under the bot name.
    // `type` number mapping from Discord:
    // 0 = Playing
    // 1 = Streaming
    // 2 = Listening
    // 3 = Watching
    // 4 = Custom
    // 5 = Competing
    activities: [
      {
        name: " ", // required by Discord API, not shown in the client
        state: "just in time",     // this is what people actually see
        type: 4,               // Custom
      },
    ],
  },
    
const { Client, GatewayIntentBits, PermissionsBitField, ActionRowBuilder, StringSelectMenuBuilder } = require('discord.js');

const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildMembers
    ]
});

// Sledování pro anti-spam
const userMessageTimestamps = new Map();

const DEFAULT_ROLES_MSG = `# role

<@&1539241329899474954> - **je vlastník serveru, který rozhoduje, co se na server přidá**

<@&1552268147179135046> - **je člen A-Teamu, který s majitelem dělá změny na serveru a hlídá ho**

<@&1539240450810978364> - **je člen A-Teamu, který může také provádět změny na serveru (na žádost majitele) a hlídá server**

<@&1552391311754133555> - **je člen A-Teamu, který testuje, zda všechno funguje**

<@&1549866576029814784> - **je role pro členy, kteří jsou aktivní na serveru a ve voice chatu (VC)**

<@&1539242304013992048> - **je role, kterou mají všichni na serveru**

<@&1543385078992871524> / <@&1543385813943984239> - **je pro členy, kteří si o roli požádají**

<@&1540028254885257256> / <@&1540028432614817862> / <@&1540028872312225914> / <@&1540028968273449002> / <@&1553041371986927657> - **je pro členy, kteří si požádají o barvu přezdívky na serveru**

<@&1553032262696566915> / <@&1553032409073586186> - **je pro členy, kteří si požádají o zobrazení svého věku**

<@&1549872649877069924> - **je pro členy, kteří dají serveru Server Boost**`;

client.once('ready', () => {
    console.log(`[BOT] Přihlášen jako: ${client.user.tag} (ID: ${client.user.id})`);
});

client.on('messageCreate', async (message) => {
    if (message.author.bot || !message.guild) return;

    if (message.member.permissions.has(PermissionsBitField.Flags.Administrator)) {
        await handleCommands(message);
        return;
    }

    const contentLower = message.content.toLowerCase();

    // 1. Anti-Link
    if (contentLower.includes("http://") || contentLower.includes("https://") || contentLower.includes("discord.gg/")) {
        try {
            await message.delete();
            const warning = await message.channel.send(`${message.author}, posílání odkazů je zakázáno!`);
            setTimeout(() => warning.delete().catch(() => {}), 5000);
            return;
        } catch (err) {
            console.error("Chyba mazání odkazu:", err);
        }
    }

    // 2. Anti-Spam
    const userId = message.author.id;
    const now = Date.now();
    if (!userMessageTimestamps.has(userId)) {
        userMessageTimestamps.set(userId, []);
    }
    let timestamps = userMessageTimestamps.get(userId);
    timestamps.push(now);
    timestamps = timestamps.filter(t => now - t <= 5000);
    userMessageTimestamps.set(userId, timestamps);

    if (timestamps.length > 5) {
        try {
            await message.delete();
            const warning = await message.channel.send(`${message.author}, nepřestávej spamovat!`);
            setTimeout(() => warning.delete().catch(() => {}), 5000);
            return;
        } catch (err) {
            console.error("Chyba mazání spamu:", err);
        }
    }

    await handleCommands(message);
});

async function handleCommands(message) {
    if (!message.content.startsWith('!')) return;

    const args = message.content.slice(1).trim().split(/ +/);
    const command = args.shift().toLowerCase();

    if (command === 'setup_roles') {
        if (!message.member.permissions.has(PermissionsBitField.Flags.Administrator)) {
            return message.reply("Na tento příkaz nemáš práva!");
        }

        try {
            // Menu pro barvy
            const rowColor = new ActionRowBuilder().addComponents(
                new StringSelectMenuBuilder()
                    .setCustomId('select_color_role')
                    .setPlaceholder('🎨 Vyber si barvu jména...')
                    .addOptions([
                        { label: 'Červená', value: '1540028254885257256', emoji: '🔴' },
                        { label: 'Modrá', value: '1540028432614817862', emoji: '🔵' },
                        { label: 'Zelená', value: '1540028872312225914', emoji: '🟢' },
                        { label: 'Fialová', value: '1540028968273449002', emoji: '🟣' },
                        { label: 'Žlutá', value: '1553041371986927657', emoji: '🟡' },
                    ])
            );

            // Menu pro věk
            const rowAge = new ActionRowBuilder().addComponents(
                new StringSelectMenuBuilder()
                    .setCustomId('select_age_role')
                    .setPlaceholder('🎂 Vyber si svůj věk...')
                    .addOptions([
                        { label: '13-17+', value: '1553032262696566915', emoji: '🔞' },
                        { label: '18+', value: '1553032409073586186', emoji: '🔞' },
                    ])
            );

            // Odeslání zprávy s textem a oběma menu
            await message.channel.send({
                content: DEFAULT_ROLES_MSG + "\n\n👇 **Vyber si své role v menu níže:**",
                components: [rowColor, rowAge]
            });

            await message.delete().catch(() => {});
        } catch (err) {
            console.error("Chyba při odesílání rolí:", err);
            message.channel.send("Nastala chyba při vytváření zprávy s rolemi.");
        }
    }
}

// Reakce na interakci s menu (barvy i věk)
client.on('interactionCreate', async (interaction) => {
    if (!interaction.isStringSelectMenu()) return;
    
    if (interaction.customId === 'select_color_role' || interaction.customId === 'select_age_role') {
        const roleId = interaction.values[0];
        const role = interaction.guild.roles.cache.get(roleId);

        if (!role) {
            return interaction.reply({ content: "⚠️ Role na serveru nebyla nalezena.", ephemeral: true });
        }

        try {
            if (interaction.member.roles.cache.has(roleId)) {
                await interaction.member.roles.remove(roleId);
                await interaction.reply({ content: `❌ Role **${role.name}** ti byla odebrána.`, ephemeral: true });
            } else {
                await interaction.member.roles.add(roleId);
                await interaction.reply({ content: `✅ Role **${role.name}** ti byla přidána!`, ephemeral: true });
            }
        } catch (err) {
            console.error(err);
            await interaction.reply({ content: "❌ Nemám oprávnění spravovat tuto roli! Ujisti se, že botova role je v seznamu rolí výše než role, kterou se snaží přidat.", ephemeral: true });
        }
    }
});

client.login(process.env.DISCORD_TOKEN);
